includes/class-wcs-failed-scheduled-action-manager.php

```diff
@@ -106,6 +114,53 @@ public function log_action_scheduler_failure( $action_id, $error ) {
 		);
 
 		update_option( WC_Subscriptions_Admin::$option_prefix . '_failed_scheduled_actions', $failed_scheduled_actions );
+
+		// If there is an exception listener and it's caught exceptions, log them for additional debugging.
+		if ( ! empty( $this->exceptions ) ) {
+			foreach ( $this->exceptions as $exception ) {
+				$message = 'Exception: ' . $exception->getMessage() . ' in ' . $exception->getFile() . ':' . $exception->getLine();
+				$this->log( $message . PHP_EOL . $exception->getTraceAsString() );
+				ActionScheduler_Logger::instance()->log( $action_id, $message );
+			}
+
+			// Now that we've logged the exceptions, we can detach the exception listener.
+			$this->clear_exceptions();
+		}
+	}
+
+	/**
+	 * Creates a new exception listener when processing subscription-related scheduled actions.
+	 *
+	 * @param int $action_id The ID of the scheduled action being ran.
+	 */
+	public function maybe_attach_exception_listener( $action_id ) {
+		$action = $this->get_action( $action_id );
+
+		if ( ! $action || ! isset( $this->tracked_scheduled_actions[ $action->get_hook() ] ) ) {
+			return;
+		}
+
+		// Add an action to detach the exception listener after the scheduled action has been executed.
+		add_action( 'action_scheduler_after_execute', [ $this, 'clear_exceptions' ] );
+		add_action( 'woocommerce_caught_exception', [ $this, 'handle_exception' ] );
```

Comment:
Do we need to `remove_action` this hook on `action_scheduler_after_execute` as well?

Something I noticed is that we're only clearing the list of exceptions once they're logged or on the  `action_scheduler_after_execute` hook, but shouldn't we clear them on the `begin` action as well? 

While it's very unlikely... what if the following scenario ever happens within the same request, won't we end up logging incorrect data?

1. Job 1 starts (successful): `action_scheduler_begin_execute` - exception listeners added
3. Job 1 finishes: `action_scheduler_after_execute` cleared
4. `woocommerce_caught_exception` is thrown by some 3rd party code and we've added it to the list of exceptions
5. Job 2 starts (failure): `action_scheduler_begin_execute`
6. `woocommerce_caught_exception`
7. Job 2 `log_action_scheduler_failure` runs and logs both exceptions from step 4 and 6 when only 6 is relevant to the scheduled action?

Reply:
yeah thats a good point. You can kind of tell from the comment above this line that is what I intended. 

```
// Add an action to detach the exception listener after the scheduled action has been executed.
```

Except I don't detach the listener at all, I just empty the exceptions. I'm not sure why I did that though. 🤔 

I'll fix that up though and retest.

Reply:
> Except I don't detach the listener at all, I just empty the exceptions. I'm not sure why I did that though. 🤔

Yeah I noticed that as well. I thought maybe your initial approach was detaching the listeners but then moved away from that before pushing the code and forgot to update the comment.

Reply:
I also confirmed your theory. Doing something like this: 


```php
ActionScheduler_QueueRunner::instance()->process_action( 51111 );
wc_caught_exception( new Exception( 'Unrelated error' ) );
ActionScheduler_QueueRunner::instance()->process_action( 51119 );
```

Where 51111 is the scheduled action ID for a successful action and 51119 fails. 

<img width="316" alt="Screenshot 2024-01-11 at 1 34 19 pm" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/8490476/1d61f936-a057-4cfc-b7be-060aa0cb5585">

You can see that because the first action doesn't detach itself, the exception thrown in between is logged for the second one.

Reply:
I've pushed https://github.com/Automattic/woocommerce-subscriptions-core/pull/552/commits/1c816bbd3a0dce2b9373b66a7bd665d13d371f84 should detach the exception listener when the action either successfully exceutes (`action_scheduler_after_execute`) or fails (`action_scheduler_failed_action` or `action_scheduler_failed_execution`)

Reply:
Thanks @james-allan that's amazing you were able to reproduce the theory! I'll give this another review/test